cli_tools/common/utils/storage/storage_client.go

```diff
@@ -145,13 +146,22 @@ func (sc *Client) DeleteGcsPath(gcsPath string) error {
 }
 
 // FindGcsFile finds a file in a GCS directory path for given file extension. File extension can
-// be a file name as well.
+// be a file name as well. The lookup is done recursively.
 func (sc *Client) FindGcsFile(gcsDirectoryPath string, fileExtension string) (*storage.ObjectHandle, error) {
-	bucketName, objectPath, err := SplitGCSPath(gcsDirectoryPath)
+	return sc.FindGcsFileDepthLimited(gcsDirectoryPath, fileExtension, -1)
+}
+
+// FindGcsFileDepthLimited finds a file in a GCS directory path for given file
+// extension up to lookupDepth deep. If lookup should be only for files directly in
+// gcsDirectoryPath, lookupDepth should be set as 0. For recursive lookup with
+// no limitations on depth, lookupDepth should be -1
+// File extension can be a file name as well.
+func (sc *Client) FindGcsFileDepthLimited(gcsDirectoryPath string, fileExtension string, lookupDepth int) (*storage.ObjectHandle, error) {
+	bucketName, lookupPath, err := SplitGCSPath(gcsDirectoryPath)
 	if err != nil {
 		return nil, err
 	}
-	it := sc.GetObjects(bucketName, objectPath)
+	it := sc.GetObjects(bucketName, lookupPath)
```

Comment:
Since you're depending on the Client object, your test code will need a lot of deep mocking.  An alternative is to express your dependency in terms of the data that you need, which, AFAICT, is an iterator-style source of strings. 

This is just an idea -- either for this PR if you like it, or a future PR where you find yourself doing deep mocking.

Reply:
What do you mean "depending on the Client object"? This is a method of storage Client.

Also, what do you mean by "test code will need a lot of deep mocking"?

Reply:
The ~20 lines of setup code here, for example:

The key insight is that the algorithmic portion of your code doesn't care about GCS or `storage.ObjectAttrs` - its main concern is filepath-like strings.
https://github.com/GoogleCloudPlatform/compute-image-tools/blob/420d6d8c0afb11d1fdba06a50c8759c4ed332524/cli_tools/common/utils/storage/storage_client_test.go#L284

Reply:
isDepthValid is tested on it's own. Then there are some more higher level tests like the one you referenced above that test FindGcsFileDepthLimited() in more detail. 

I don't quite see what is your concern here.